### PR TITLE
Allow video url redirects

### PR DESCRIPTION
**Pull request by luoser**
## what
Video playback has been broken on our HQ environments with a 302 redirect error since our HQ video urls redirect.

Luckily the fix is nice: [ExoPlayer has multiple constructors for `DefaultUriDataSource`](http://google.github.io/ExoPlayer/doc/reference-v1/com/google/android/exoplayer/upstream/DefaultUriDataSource.html) (which we use in our `KSRendererBuilder`) one of which allows us to set a boolean for `allowCrossProtocolRedirects`. You go ExoPlayer.

Bonus modernization of our #weird `VideoViewModel` and refactoring of our `ProjectViewModel` and tests.

### CODE REVIEW

File: app/src/main/java/com/kickstarter/ui/activities/VideoActivity.java [hunk 1]
```diff
@@ -44,7 +43,7 @@ public void onCreate(final @Nullable Bundle savedInstanceState) {
     setContentView(R.layout.video_player_layout);
     ButterKnife.bind(this);
 
-    viewModel.outputs.video()
+    viewModel.outputs.preparePlayerWithUrl()
```
Comment (mbrandonw) [inline app/src/main/java/com/kickstarter/ui/activities/VideoActivity.java]: muuuuch better name

File: app/src/main/java/com/kickstarter/viewmodels/ProjectViewModel.java [hunk 2]
```diff
@@ -172,21 +172,21 @@ public ViewModel(final @NonNull Environment environment) {
...
       this.startProjectUpdatesActivity = currentProject.compose(takeWhen(this.updatesTextViewClicked));
+      this.startVideoActivity = currentProject.compose(takeWhen(this.playVideoButtonClicked));
       this.startViewPledgeActivity = currentProject.compose(takeWhen(this.viewPledgeButtonClicked));
 
       this.shareButtonClicked
         .compose(bindToLifecycle())
         .subscribe(__ -> this.koala.trackShowProjectShareSheet());
 
-      this.playVideo
+      this.startVideoActivity
```
Comment (mbrandonw) [inline app/src/main/java/com/kickstarter/viewmodels/ProjectViewModel.java]: muuuuuch better name

File: app/src/main/java/com/kickstarter/viewmodels/VideoViewModel.java [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+import com.kickstarter.ui.activities.VideoActivity;
+
+import rx.Observable;
+import rx.subjects.BehaviorSubject;
+
+public interface VideoViewModel {
+
+  interface Outputs {
+    /** Emits the video for the player. */
+    Observable<String> preparePlayerWithUrl();
```
Comment (mbrandonw) [inline app/src/main/java/com/kickstarter/viewmodels/VideoViewModel.java]: looooovin these new modern vm's

File: app/src/main/java/com/kickstarter/libs/KSRendererBuilder.java [hunk 4]
```diff
@@ -51,7 +51,7 @@ public KSRendererBuilder(final @NonNull Context context, final @NonNull String v
   @Override
   public void buildRenderers(final @NonNull KSVideoPlayer player) {
     final Allocator allocator = new DefaultAllocator(BUFFER_SEGMENT_SIZE);
-    final DataSource dataSource = new DefaultUriDataSource(context, videoLink);
+    final DataSource dataSource = new DefaultUriDataSource(context, null, videoLink, true);
```
Comment (mbrandonw) [inline app/src/main/java/com/kickstarter/libs/KSRendererBuilder.java]: so this is the flag that allows redirects?

def makes you miss swift's named arguments when reading this code......

Comment (luoser) [inline app/src/main/java/com/kickstarter/libs/KSRendererBuilder.java]: yep

**Pull request merged**